edgepi-dac.js

```diff
@@ -0,0 +1,72 @@
+const { checkValidWriteInput } = require('./helpers/nodeHelpers');
+
+module.exports = function (RED) {
+    const rpc = require('@edgepi-cloud/edgepi-rpc')
+  
+    function DacNode(config) {
+      // Create new node instance with user config
+      RED.nodes.createNode(this, config);
+      const node = this;
+      const ipc_transport = "ipc:///tmp/edgepi.pipe"
+      const tcp_transport = `tcp://${config.tcpAddress}:${config.tcpPort}`
+      const transport = (config.transport === "Network") ? tcp_transport : ipc_transport;
+      const voltage = parseFloat(config.voltage);
+      const channel = rpc.DACChannel[config.dacChannel];
+
+      // Init dac
+      const dac = new rpc.DacService(transport)
+  
+      if (dac){
+        console.info("DAC node initialized on:", transport);
+        node.status({fill:"green", shape:"ring", text:"dac initialized"});
+      }
+
+      // Enforce gain on write method and check valid configuration input
+      let validWriteInput;
+      if(config.method === "write"){
+        dac.setDacGain(true,true);
+        validWriteInput = checkValidWriteInput(voltage, node);
+      }
+
+      
+  
+      // Input event listener
+      node.on('input', async function(msg,send,done){
+        if(config.method === "write" && !validWriteInput){
+          node.status({fill:"red", shape:"dot", text:"node made no call because it has invalid input."});
+          return;
+        }
+
+        node.status({fill:"green", shape:"dot", text:"input recieved"});
+        try{
+          let response;
+          if(config.method === "read"){
+            const {voltageVal} = await dac.getState({
+              analogOut: channel,
+              voltage: true
+            })
+            response = voltageVal;
+          }
+          else if (config.method === "write"){
+            response = await dac.writeVoltage(channel, voltage);
+          }
+          msg.payload = response;
+        }
```

Comment:
i think it's good to have a final `else` block here as like a fallback. so if you get an unexpected value for `config.method` it goes there

```
else:
    throw new Error(`Unexpected value: ${config.method}`);
```

Reply:
It's unlikely that will happen since the method options are hard coded into the UI as a dropdown menu but if there is an unexpected value it should be caught in the `try` `catch` block that surrounds this.

Reply:
how does the try-catch catch unexpected values? i don't see it in the code

if you're absolutely sure there's only 2 options then it's fine

Reply:
oh wait I was thinking the wrong thing, you're right it won't catch anything. I added the fallback in case new methods are added -> 631a31d0b183d11999d54c3ce5713924f58a5b4d